[WIP] Generic memory block allocator

tobhe opened the pull request:
This PR adds a generic block allocator, allowing "dynamic" allocations of structures of the same size in a static memory pool.
It is meant to replace malloc in cases where only fixed size blocks are allocated and freed (see #7615).

### Example usage
```
// Initialize static memory pool "mem" which can hold 3 somestruct_t
MEMARRAY(mem, somestruct_t, 3);

// Initialize mem as a linked list of free elements
memarray_init(&mem);

// Allocate structure
mem = memarray_alloc(&mem);

// Do whatever needs to be done
...

// Free struct
memarray_free(&mem, mem);
```

### Work in Progress
Things to be done:
- Add documentation
- Add unittests

smlng commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,55 @@
+#endif
+
+/**
+ * @defgroup    sys_memarray memory array allocator
```
please move this comment block up, directly below the copy right statement - just for consistency with RIOT codebase.

smlng commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,55 @@
+#ifndef MEMARRAY_H
+#define MEMARRAY_H
+
+#include "stdint.h"
```
these standard includes should go into `<` and `>`

smlng commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,55 @@
+ * @brief       pseudo dynamic allocation in static memory arrays
+ * @author      Tobias Heider <heidert@nm.ifi.lmu.de>
+ */
+typedef struct {
```
struct need documentation, plus its elements, i.e. 
```
    size_t size;    /**< size of memory array */
    size_t count;   /**< number of ... */
```

smlng commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,55 @@
+                            .first_free = _data_##name,                        \
+                            .data = _data_##name};
+
+void memarray_init(memarray_t *mem);
```
add full doxygen style documentation for all functions here and below

miri64 commented:
How is this different from `tlsf`? What are it's benefits, what are its drawbacks compared to it?

kaspar030 commented:
While sth like this useful, I think the implementation has several drawbacks, mostly in the way free blocks are handled.
In the current implementation, every free iterates through all free blocks.

How about this:
- assume the smallest supported blocksize will be 4 bytes (imo it currently does anyways?)
- on initialization, initialize the memory divided into blocks of the desired size as a clist. Use the first four bytes of each block as list entry, and link all the blocks
- reduce the memarray struct to contain a clist entry to the first block
- use simple clist_lpop() / clist_rpush() for alloc/free

That way the actual struct would be down to 4 bytes, and both allocating/deallocating can be done in O(1), also re-using existing code.

tobhe commented on sys/memarray/memarray.c:
```diff
@@ -0,0 +1,43 @@
+    }
+}
+
+uint8_t memarray_free(memarray_t *mem, void *ptr)
```
right, i think to achieve O(1) it would be necessary to either not check boundaries at all and just assume ptr is in the memory pool, or to check whether `ptr >= mem->data` and `ptr < mem->data+mem->count*mem->size` which doesn't check whether the pointer is aligned in the pool.

tobhe commented:
@kaspar030 
I think you are absolutely right, i will look into your solution. The current implementation does indeed assume a minimum blocksize of 4 to store pointers to the next free element.

tobhe commented:
> How is this different from tlsf? What are it's benefits, what are its drawbacks compared to it?

@miri64 TLSF is meant to be a full on malloc replacement, allowing allocations of variable size and thus is way bigger and more complex. This module is meant to solve a simpler problem, where for each pool, the size of blocks as well as the number of blocks in a pool are known at compile time. 

An example use case would be a network protocol package (like tinydtls), which needs to allocate one structure (holding control information as the peers address etc.) per open connection. Using TLSF would be overkill, as it would require including an external packet which provides a lot of functionality that is not really needed here.
As the size of the structure as well as the maximum required number of connections are usually known at compile time, a static array of structures, which is managed with a linked list (either the way it currently works or the way @kaspar030 proposed), can provide the same performance of O(1) allocations and frees, with less fragmentation, smaller code size and without the need for an external package.

gebart commented:
Since the memarray is a coherent array of blocks, would it be possible to use a bit map of the free blocks, instead of a linked list? I believe this would make free O(1) and malloc O(N) with lower memory overhead than the linked list implementation.

tobhe commented:
> Since the memarray is a coherent array of blocks, would it be possible to use a bit map of the free blocks, instead of a linked list? I believe this would make free O(1) and malloc O(N) with lower memory overhead than the linked list implementation.

I've thought about using a bitmap index, i think it is even possible to malloc in O(1) using bitmagic. However I think the allocator list is the least complex solution and does not have notably more overhead than the bitmap in small pools and scales better.

> That way the actual struct would be down to 4 bytes, and both allocating/deallocating can be done in O(1), also re-using existing code.

I've thought about what you said and I actually don't see any real advantage of using a clist. The non-clist approach works just as well as the one you described. The only reason `memarray_free` is O(N) is the check whether the pointer is actually a element of the list (which I think is actually not necessary as no other free implementation does this).  Other than that, size and number of elements are only needed in the `memarray_init` function, regardless whether a clist or a normal list is initialized, so the memarray_t could be reduced to one 4 byte "head of the list" pointer in both cases.

kaspar030 commented on sys/memarray/memarray.c:
```diff
@@ -0,0 +1,34 @@
+ */
+
+#include "memarray.h"
+#include <string.h>
```
please reverse header sequence

tobhe replied:
done

kaspar030 commented:
> I actually don't see any real advantage of using a clist.

I agree, it looks much better now and using clist would complicate things.

I'm wondering about the use of ```memcpy``` and your assumptions on alignment. It looks like there's no alignment being done, and that's why you actually use ```memcpy``` for storing/retrieving the pointers, right?

tobhe commented:
> I'm wondering about the use of `memcpy` and your assumptions on alignment. It looks like there's no alignment being done, and that's why you actually use `memcpy` for storing/retrieving the pointers, right?

Indeed, because `memcpy` allows me to not care about the alignment.

tobhe commented:
Small update: 
@rfuentess contributed a very nice testing application,
also @smlng: I guess the coding style should be fine now.

tobhe commented on tests/memarray/README.md:
```diff
@@ -0,0 +1,18 @@
+At the beginning of the tests is printed the memory being used by the threads.
+At the end of the tests, the threads memory usage is printed once more time.
+
+This test is passed if the memory used by the threads are correct and if the legend
```
thanks, fixed it

smlng commented on tests/memarray/Makefile:
```diff
@@ -0,0 +1,16 @@
+# Comment this out to disable code in RIOT that does safety checking
+# which is not needed in a production environment but helps in the
+# development process:
+CFLAGS += -DDEVELHELP
```
please remove, not needed since #8080

smlng commented on tests/memarray/README.md:
```diff
@@ -0,0 +1,18 @@
+Expected result
+===============
+
+This application should runs a number of tests equal to NUMBER_OF_TESTS (Default 12).
```
typo /runs/run/ or remove `should`

smlng commented on tests/memarray/README.md:
```diff
@@ -0,0 +1,18 @@
+
+This application should runs a number of tests equal to NUMBER_OF_TESTS (Default 12).
+
+At the beginning of the tests is printed the memory being used by the threads.
```
wording, consider `At the beginning of the tests memory usage of each thread is printed.`

smlng commented on tests/memarray/README.md:
```diff
@@ -0,0 +1,18 @@
+Background
+==========
+
+The module `memarray` is the static memory allocator for RIOT O.S.
```
typo: /RIOT O.S./RIOT-OS/

smlng commented on tests/memarray/main.c:
```diff
@@ -0,0 +1,121 @@
+
+#include "memarray.h"
+
+#define MAX_NUMBER_BLOCKS 10
```
align indention of defines, add parentheses and consider making the unsigned, i.e. like 
```
#define MAX_NUMBER_BLOCKS   (10U)
#define MESSAGE_SIZE        (8U)
#define NUMBER_OF_TESTS     (12U)
```

smlng commented on tests/memarray/main.c:
```diff
@@ -0,0 +1,121 @@
+
+MEMARRAY(block_storage, sizeof(struct block_t), MAX_NUMBER_BLOCKS)
+
+
```
remove double empty line

smlng commented on tests/memarray/main.c:
```diff
@@ -0,0 +1,121 @@
+    struct block_t *old;
+
+    while (head) {
+
```
remote empty line

smlng commented on tests/memarray/main.c:
```diff
@@ -0,0 +1,121 @@
+
+int main(void)
+{
+
```
remove empty line

smlng commented on sys/memarray/memarray.c:
```diff
@@ -0,0 +1,34 @@
+
+void memarray_init(memarray_t *mem, size_t size, size_t num)
+{
+    for (size_t i = 0; i < num - 1; i++) {
```
put parentheses: `(num - 1)`

smlng commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,76 @@
+} memarray_t;
+
+/**
+ * @brief Define static memory pool
```
add empty line below

smlng commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,76 @@
+    static memarray_t name = { .first_free = _data_ ## name };
+
+/**
+ * @brief Initialize memarray pool with free list
```
dito, add empty line between `@brief` and `@param`s, here and below

smlng commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,76 @@
+/**
+ * @brief Allocate memory chunk in memarray pool
+ * @param[in,out]  mem      memarray pool to allocate block in
+ * @return pointer to newly allocated memarray chunk
```
mhm, I'm unsure if that's correct usage of `@return`, bc the function is `void` but uses a param as return.

tobhe commented:
@smlng I think i fixed it all

cladmi commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,80 @@
+ * @param[in]   size    size of single memory chunk
+ * @param[in]   num     number of chunks in pool
+ */
+void memarray_init(memarray_t *mem, size_t num, size_t size);
```
By removing num/size from `memarray_t`, the api now requires repeating them for init, which requires making them public if different modules allocate and init it.

Also, it makes `free` unsafe for invalid pointers, it may always be the case on free but few asserts could help debug mistakes.

tobhe replied:
> By removing num/size from memarray_t, the api now requires repeating them for init, which requires making them public if different modules allocate and init it.

@all any other opinions on this? otherwise i will just put them back into the memarray_t

>  Also, it makes free unsafe for invalid pointers, it may always be the case on free but few asserts could help debug mistakes.

i don't know about any free where this is not the case, but we can assert the pointer to be between `_data` and `_data + (num*size)`, maybe also the alignment

obgm commented:
Would be great if this was resolved as the ongoing work on the tinydtls port (see  PR #7615) is blocking on this.

tobhe commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,80 @@
+ * @param[in]  num     number of chunks in pool
+ */
+#define MEMARRAY(name, size, num)                                         \
+    static uint8_t _data_ ## name[num * size];                            \
```
> Why change from structure to size here ?

I'm not sure why i changed this in the first place. I think i will change it back.

> You silently expect size >= sizeof(void *), I would just allocate big enough cells all the time.
> By principle, I think its ok to allocate '8' length items even if requester asks for '5'.

agree, thanks for pointing this out

tobhe commented:
@obgm Thanks for the reminder, I've been quite busy but i'll try to spare some time to fix this

tobhe commented:
@cladmi: I changed the API according to your recommendations. Also alignment and the minimum size is now `sizeof(void*)` as elements of the array are now unions of `structure` and `void*`.
Additionally I added some debug outputs based on what @rfuentess said he found useful.

obgm commented on sys/include/memarray.h:
```diff
@@ -29,19 +29,27 @@ extern "C" {
-    static uint8_t _data_ ## name[num * size];                            \
-    static memarray_t name = { .first_free = _data_ ## name };
+#define MEMARRAY(name, structure, numb)                                           \
+    union el {                                                                    \
```
One issue with the declaration of `union el` here is that you now cannot invoke `MEMARRAY` twice:

```shell
n file included from /home/bergmann/src/RIOT/examples/dtls-echo/bin/pkg/native/tinydtls/crypto.c:122:0:
/home/bergmann/src/RIOT/sys/include/memarray.h:45:11: error: redefinition of ‘union el’
     union el {                                                                    \
           ^
/home/bergmann/src/RIOT/examples/dtls-echo/bin/pkg/native/tinydtls/crypto.c:124:1: note: in expansion of macro ‘MEMARRAY’
 MEMARRAY(security_storage,  dtls_security_parameters_t, DTLS_SECURITY_MAX)
 ^~~~~~~~
/home/bergmann/src/RIOT/sys/include/memarray.h:45:11: note: originally defined here
     union el {                                                                    \
           ^
/home/bergmann/src/RIOT/examples/dtls-echo/bin/pkg/native/tinydtls/crypto.c:123:1: note: in expansion of macro ‘MEMARRAY’
 MEMARRAY(handshake_storage, dtls_handshake_parameters_t, DTLS_HANDSHAKE_MAX)
 ^~~~~~~~
```

tobhe replied:
@obgm: absolutely right, thanks for noticing. I'm gonna append `name` as i did with `_data ## name`

aabadie commented on sys/Makefile:
```diff
@@ -121,6 +121,9 @@ endif
     DIRS += net/link_layer/l2filter
 endif
+ifneq (,$(filter memarray,$(USEMODULE)))
+    DIRS += memarray
```
Indent should be 2 spaces here (RIOT style for Makefiles)

tobhe replied:
fixed, thanks

aabadie commented on tests/memarray/main.c:
```diff
@@ -0,0 +1,118 @@
+        }
+        else {
+            memarray_free(&block_storage, head);
+            // printf("\tLast block released, *head: %p\n", (void *) head);
```
I guess this line could be removed ?

tobhe replied:
fixed

smlng commented:
untested ACK

smlng commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,88 @@
+/**
+ * @brief Initialize memarray pool with free list
+ *
+ * @param[out]  mem     memarray pool to initialize
```
indention looks off, compared to below

kaspar030 commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,88 @@
+ * @param[in]  structure  structure in single memory chunk
+ * @param[in]  numb       number of chunks in pool
+ */
+#define MEMARRAY(name, structure, numb)                                           \
```
Please get rid of the macro.

tobhe replied:
I agree that macros should be avoided, but I can not think of a better way to build an abstraction which declares the data array and memarray struct with a unique name and thus can be used to declare more than one such array. The only other option I see is letting the user do the declaration, which I think makes it more complicated to use. What do you think would be the best alternative?

OlegHahm commented:
Sorry to hop in that late. Since I've played around with several memory pool allocators in the past, I just wondered if you compared your implementation with any existing ones?

kaspar030 commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,88 @@
+ * @brief Initialize memarray pool with free list
+ *
+ * @param[out]  mem     memarray pool to initialize
+ * @param[in]   size    size of single memory chunk
```
there'd documentation for non-existing parameters

kaspar030 commented:
> I agree that macros should be avoided, but I can not think of a better way to build an abstraction which declares the data array and memarray struct with a unique name and thus can be used to declare more than one such array. The only other option I see is letting the user do the declaration, which I think makes it more complicated to use. What do you think would be the best alternative?

I think it is totally acceptable to have the user declare and supply a buffer. It is definitely preferable to that macro construct.

tobhe commented:
@OlegHahm: 
> Since I've played around with several memory pool allocators in the past, I just wondered if you compared your implementation with any existing ones?

The current implementation surely is a very naïve one compared to other existing implementations, but I think that even in it's current non optimized form it surely is an improvement to using `malloc` (especially when there is no `free`). As @cladmi recommended, my plan is currently to coverge the API first and then eventually improve the implementation in another PR. 
That being said: If you know of any useful sources regarding more efficient implementations which I should have a look into, I would be happy to do so.

@kaspar030:
> I think it is totally acceptable to have the user declare and supply a buffer. It is definitely preferable to that macro construct.

Thanks for sharing, then I will do exactly this

kaspar030 commented on tests/memarray/Makefile:
```diff
@@ -0,0 +1,11 @@
+# name of your application
```
first two lines can be removed, they're set in the include below

kaspar030 commented on tests/memarray/main.c:
```diff
@@ -0,0 +1,122 @@
+
+void free_memory(struct block_t *head)
+{
+
```
extra empty line

kaspar030 commented:
> Because of memory efficiency i use the pointer to the head of the free list as a initial pointer to the memory location, which seems quite dirty, but should not make problems as this is the same what implicitly happend before. If you think this is too hacky, let me know and i will add seperate data and first_free fields to the struct.

Hm, how about changing the memarray_init() signature back to include the necessary info (memarray_t*, ptr, element size, num of elements)?

That way the initialization would  look more standard, like:
```
my_type_t _array[NUM_ELEMENTS];
memarray_t memarray;
memarray_init(&memarray, _array, sizeof(_array[0]), NUM_ELEMENTS);
```

Or maybe ```memarray_init(..., ptr, total_size, element_size)``` is more intuitive?

tobhe commented:
> Hm, how about changing the memarray_init() signature back to include the necessary info (memarray_t*, ptr, element size, num of elements)?

I think you're right, this seems to be the better solution. As far as I've seen other modules, e.G. [crypto/ciphers](https://riot-os.org/api/group__sys__crypto.html) do their initialization in a similar way, so it would also add to the consistency of the API as a whole.

tobhe commented:
@cladmi @kaspar030 
I have changed the api as requested, also i think i did get the documentation right this time.
any more comments on the current version?

bergzand commented:
@tobhe ping

I'd like to use this in combination with the cn-cbor package :)

tobhe commented:
@bergzand As I am back from vaction now and just merged @cladmi's format string fixes I think this might be ready to get merged soon™. I was not planning to make any bigger API changes, and overall everyone seems happy with the current state (or at least no one has raised any concerns). So, feel free to use this, I'm glad there is so much interest in this

bergzand commented:
> As I am back from vaction now

Didn't want to rush you here, thanks for continuing this. :)

@kaspar030, @cladmi: Are all your comments addressed? Is this ready to squash?

tobhe commented:
@miri64 thanks, i totally forgot the @pre. I think i got them all ;)

miri64 commented on tests/memarray/main.c:
```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (C) 2017 Inria
```
2018 Tobias Heider <heidert@nm.ifi.lmu.de> ;-)

tobhe replied:
That one was initially done by @rfuentess so I don't want to change it without his consent. 
I can update the date though.

rfuentess replied:
For me is OK to share authorship or given it to @tobhe, after all he had to upgrade the code.

miri64 commented on sys/include/memarray.h:
```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2017 Tobias Heider <heidert@nm.ifi.lmu.de>
```
2017-18 ;-)

miri64 commented on sys/memarray/memarray.c:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2017 Tobias Heider <heidert@nm.ifi.lmu.de>
```
2017-18 ;-)

miri64 commented:
I'm fine, after the copyright issue is solved.

kaspar030 commented on tests/memarray/README.md:
```diff
@@ -0,0 +1,18 @@
+Background
+==========
+
+The module `memarray` is the static memory allocator for RIOT-OS
```
IMO this needs re-phrasing. What is a "static memory allocator"?

cladmi replied:
What do you think of "fixed-size blocks allocator"?

Taken from https://en.wikipedia.org/wiki/Memory_management#FIXED-SIZE and I find it clearer than "memory pool allocator"

tobhe replied:
I think "fixed-size block allocator" and "memory pool allocator" both are true, but highlight different properties of the allocator. Also, the fact that it just allocates fixed-size blocks may be the more distinct feature of this implementation, so I'm on your side with this.

kaspar030 commented on sys/memarray/memarray.c:
```diff
@@ -0,0 +1,52 @@
+    mem->size = size;
+    mem->num = num;
+
+    for (size_t i = 0; i < (mem->num - 1); i++) {
```
This makes ```mem->num - 1``` re-evaluate on every loop iteration.

cladmi replied:
Does it ? has `mem->num` is not volatile there is no reason for the compiler to re-read it I think.
When replacing with an intermediate variable I get the same size for `samr21-xpro`, `wsn430-v1_3b` and `native`.
If you have another solution that reduces code size it would be good.

When using `(num -1)` in the loop it gets bigger for `samr21-xpro`, no idea why.

kaspar030 replied:
Any function call in between two dereferences usually forces re-read from memory. The called function might have changed what's behind that pointer, after all. But maybe the compiler has outsmarted me (again).

cladmi replied:
Yeah `memcpy` could be doing some stuff to `mem->num` by overwriting it if `mem` and `data` overlap and in that case the compiler should indeed re-read it. Maybe it is re-read but does not cost more.


I found one that reduces code size and also solves the `num == 0` case.
But it would print many debug free message at init so maybe not that good.

```
    for (char *cur = data; cur < (char *)data + (num * size); cur += size) {
        memarray_free(mem, cur);
    }
```
We can keep the optimizations to another PR this way it is not tied to tobhe.

cladmi commented on sys/memarray/memarray.c:
```diff
@@ -0,0 +1,52 @@
+
+void memarray_init(memarray_t *mem, void *data, size_t size, size_t num)
+{
+    assert((mem != NULL) && (data != NULL) && (size >= sizeof(void *)));
```
In the same part of having assert for the API I also saw that there could be an assert for `num != 0`  as its a case that is not handled right now.

tobhe replied:
True, thanks for the hint

bergzand commented:
:tada:

tobhe commented:
@cladmi You're welcome, thanks for all the help

aabadie commented:
Nice, this will certainly interest @rfuentess for tinydtls

miri64 commented:
:tada: @tobhe congratz!
